### ISSUE TITLE
cmd/compile: escape analysis oddity

### ISSUE THREAD

**Issue by davecheney** [golang/go#7213]
<pre>What steps will reproduce the problem?

package q

import &quot;io&quot;

type Thing struct {
        r io.Reader
}

func (t *Thing) Read() error {
        _, err := t.r.Read(nil)
        return err
}

func Use() error {
        var t Thing
        return t.Read()
}

% go build -gcflags -m leak.go
# command-line-arguments
./leak.go:9: leaking param: t
./leak.go:15: moved to heap: t
./leak.go:16: t escapes to heap

What is the expected output? What do you see instead?

Why does t leak on like 9, what is it about referencing t.r that causes t to leak onto
the heap.

Please use labels and text to provide additional information.

This question arose from an investigation if json.NewDecoder could be stack allocated.</pre>

**Comment by freeformz** [comment]
FWIW: This is the output under go1.7rc2

```console
$ go build -gcflags '-m -l' leak.go
# command-line-arguments
./leak.go:9: leaking param content: t
./leak.go:16: Use t does not escape
```

While t is detected as leaking, it does not escape.

**Comment by cespare** [comment]
Yeah, seems like this minimal repro is no longer a repro, yet the &json.Decoder literal from json.NewDecoder still escapes (even though json.NewDecoder is inlined).

(Go 1.8.)

**Comment by mdempsky** [comment]
The original minimized repro case has been fixed (since Go 1.7, evidently), so I'm going to close this as fixed.
